appservice/src/deploy/localGitDeploy.ts

```diff
@@ -45,17 +65,29 @@ export async function localGitDeploy(client: SiteClient, fsPath: string): Promis
             }
             return undefined;
             // tslint:disable-next-line:no-unsafe-any
-        } else if (err.message.indexOf('error: failed to push') >= 0) {
-            const forcePush: vscode.MessageItem = { title: localize('forcePush', 'Force Push') };
-            const pushReject: string = localize('localGitPush', 'Push rejected due to Git history diverging.');
-            await ext.ui.showWarningMessage(pushReject, forcePush, DialogResponses.cancel);
-            // tslint:disable-next-line:no-floating-promises
-            localGit.push(remote, 'HEAD:master', { '-f': true });
         } else {
             throw err;
         }
     }
 
-    ext.outputChannel.appendLine(formatDeployLog(client, (localize('localGitDeploy', `Deploying Local Git repository to "${client.fullName}"...`))));
-    await waitForDeploymentToComplete(client, kuduClient, commitId);
+    async function pushAndWaitForDeploymentToComplete(token?: vscode.CancellationToken, forcePush: boolean = false): Promise<void> {
+        await new Promise((resolve: () => void, reject: (error: Error) => void): void => {
+
+            // for whatever reason, is '-f' exists, true or false, it still force pushes
+            const pushOptions: git.Options = forcePush ? {'-f': true} : {};
+
+            localGit.push(remote, 'HEAD:master', pushOptions).catch(async (error) => {
+                tokenSource.cancel();
+                // tslint:disable-next-line:no-unsafe-any
+                reject(error);
+            });
+
+            waitForDeploymentToComplete(client, kuduClient, commitId, token).then(resolve).catch((error: Error) => {
+                // waitForDeploymentToComplete throws a UserCancelledError when cancelled so that we don't mistake it as "completing"
+                if (!parseError(error).isUserCancelledError) {
```

Comment:
I think it's a little confusing for `waitForDeploymentToComplete` to throw a _User_CancelledError. What if it just returns nothing and you do something like this:
```
waitForDeploymentToComplete(client, kuduClient, commitId, token).then(() => {
    if (!tokenSource.canceled) {
        resolve();
    }
}).catch(reject);
```

Reply:
Actually if you take my above suggestion to return nothing and if you switch these lines:
```
// tslint:disable-next-line:no-unsafe-any
reject(error);
tokenSource.cancel();                
```

You can just do this:
```
waitForDeploymentToComplete(client, kuduClient, commitId, token).then(resolve).catch(reject);
```

Since `reject` is called before `cancel`, you don't have to worry about the result of `waitForDeploymentToComplete` since the promise was guaranteed to already have finished.

Reply:
Doesn't this mean that tokenSource will never be cancelled? Since it would reject with the error before it can cancel the token.

Reply:
You're thinking of `throw error;`, which cancels the current code path. However, `reject(error);` is different and should continue

Reply:
ping

Reply:
Ahhh, okay, that makes a lot more sense. 👍